cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu

```diff
@@ -0,0 +1,238 @@
+{
+  const cuda::std::size_t num_items = 1ull << 33;
+  thrust::device_vector<cuda::std::size_t> out_num_runs(1, -1);
+
+  // Note intentionally no discard_iterator as we want to ensure nothing is written to the output arrays
+  run_length_encode(thrust::make_counting_iterator(cuda::std::size_t{0}),
+                    static_cast<cuda::std::size_t*>(nullptr),
+                    static_cast<cuda::std::size_t*>(nullptr),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_num_runs.front() == 0);
+}
+#endif
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle different counting types", "[device][run_length_encode]")
+{
+  const int num_items = 1;
+  thrust::device_vector<int> in(num_items, 42);
+  thrust::device_vector<int> out_num_runs(1, 42);
+
+  // Note intentionally no discard_iterator as we want to ensure nothing is written to the output arrays
+  run_length_encode(in.begin(),
+                    static_cast<cuda::std::size_t*>(nullptr),
+                    static_cast<std::uint16_t*>(nullptr),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_num_runs.front() == 0);
+}
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle all unique", "[device][run_length_encode]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = 10;
+  thrust::device_vector<int> out_num_runs(1, -1);
+
+  run_length_encode(thrust::make_counting_iterator(type{}),
+                    static_cast<int*>(nullptr),
+                    static_cast<int*>(nullptr),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_num_runs.front() == 0);
+}
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle all equal", "[device][run_length_encode]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = 10;
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<int>  out_offsets(1, -1);
+  thrust::device_vector<int>  out_lengths(1, -1);
+  thrust::device_vector<int>  out_num_runs(1, -1);
+  c2h::gen(CUB_SEED(2), in);
+  thrust::fill(in.begin(), in.end(), in.front());
+
+  run_length_encode(in.begin(),
+                    out_offsets.begin(),
+                    out_lengths.begin(),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_offsets.front()  == 0);
+  REQUIRE(out_lengths.front()  == num_items);
+  REQUIRE(out_num_runs.front() == 1);
+}
+
+template<class T, class Index>
+bool validate_results(const thrust::device_vector<T>&     in,
+                      const thrust::device_vector<Index>& out_offsets,
+                      const thrust::device_vector<Index>& out_lengths,
+                      const thrust::device_vector<Index>& out_num_runs,
+                      const int num_items)
+{
+  for(cuda::std::size_t run = 0; run < static_cast<cuda::std::size_t>(out_num_runs.front()); ++run) {
```

Comment:
hint: if the comparison expression is not optimized, it'd be a memory transfer on each iteration to fetch num runs.

Reply:
But I can rely on the optimizer cant I?

... Moves code out of loop ....

Reply:
Done